modules/caddyhttp/reverseproxy/reverseproxy.go

```diff
@@ -614,7 +614,12 @@ func (h *Handler) reverseProxy(rw http.ResponseWriter, req *http.Request, di Dia
 	// some apps need the response headers before starting to stream content with http2,
 	// so it's important to explicitly flush the headers to the client before streaming the data.
 	// (see https://github.com/caddyserver/caddy/issues/3556 for use case)
-	if req.ProtoMajor == 2 && res.ContentLength == -1 {
+	// we have to check encoding here, only flush headers with identity encoding.
+	// Non-identity encoding might combine with "encode gzip" directive,
+	// in that case, if body size larger than enc.MinLength,
+	// upper level encode handle might have Content-Encoding header to write.
+	// (see https://github.com/caddyserver/caddy/issues/3606 for use case)
+	if req.ProtoMajor == 2 && res.ContentLength == -1 && req.Header.Get("Accept-Encoding") == "identity" {
```

Comment:
If Accept-Encoding is empty (not specified by the client), would that also be permissible for flushing?

Reply:
According to [rfc2616-sec14](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html), it's permissible.
Should we consider dealing with the following form? It means accepting all but "identity".

```
Accept-Encoding: identity;q=0, *;q=1
```